pkg/guacanalytics/patchPlanning.go

```diff
@@ -0,0 +1,179 @@
+//
+// Copyright 2023 The GUAC Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package guacanalytics
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Khan/genqlient/graphql"
+	model "github.com/guacsec/guac/pkg/assembler/clients/generated"
+	"github.com/guacsec/guac/pkg/misc/depversion"
+)
+
+type DfsNode struct {
+	expanded bool // true once all node neighbors are added to queue
+	Parent   string
+	depth    int
+	nodeType string // packageName, packageVersion
+}
+
+type queueValues struct {
+	path    []string
+	nodeMap map[string]DfsNode
+	now     string
+	nowNode DfsNode
+	queue   []string
+}
+
+// TODO: make more robust using predicates
+func SearchDependenciesFromStartNode(ctx context.Context, gqlclient graphql.Client, startID string, stopID string, startType string, maxDepth int) (map[string]DfsNode, error) {
```

Comment:
if stopID is optional it should be a pointer instead to indicate if it's required

Reply:
Fixed (changed in tests as well).